src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1022,250 +1015,222 @@ private ImmutableArray<BoundNode> ConvertCollectionExpressionElements(
+                                    item,
+                                    arg.elementConversion,
+                                    arg.elementType,
+                                    arg.elementKeyValueTypes,
+                                    arg.useIndexer,
+                                    implicitReceiver,
+                                    arg.collectionInitializerAddMethodBinder,
+                                    diagnostics);
+                            },
+                            (elementType, useIndexer, elementConversion, elementKeyValueTypes, collectionInitializerAddMethodBinder),
+                            diagnostics);
+                        break;
+                    case BoundKeyValuePairElement keyValuePairElement:
+                        convertedElement = convertKeyValuePair(
+                            this,
+                            keyValuePairElement,
+                            elementConversion,
+                            elementType,
+                            elementKeyValueTypes,
+                            useIndexer,
+                            implicitReceiver,
+                            collectionInitializerAddMethodBinder,
+                            diagnostics);
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(element);
                 }
+                builder.Add(convertedElement);
             }
-            else if (elementType is { } && ConversionsBase.TryGetCollectionKeyValuePairTypes(Compilation, collectionTypeKind, elementType) is (var elementKeyType, var elementValueType))
-            {
-                var elementConversions = conversion.UnderlyingConversions;
 
-                Debug.Assert(elementType is { });
-                Debug.Assert(elementConversions.All(c => c.Exists));
-                Debug.Assert(implicitReceiver is { });
+            Debug.Assert(conversionIndex == elementConversions.Length);
+            conversion.MarkUnderlyingConversionsChecked();
 
-                int conversionIndex = 0;
-                foreach (var element in elements)
+            return builder.ToImmutableAndFree();
+
+            static BoundNode convertKeyValuePair(
+                Binder binder,
+                BoundKeyValuePairElement keyValuePairElement,
+                Conversion elementConversion,
+                TypeSymbol elementType,
+                (TypeSymbol Key, TypeSymbol Value)? elementKeyValueTypes,
+                bool useIndexer,
+                BoundObjectOrCollectionValuePlaceholder? implicitReceiver,
+                CollectionInitializerAddMethodBinder? collectionInitializerAddMethodBinder,
+                BindingDiagnosticBag diagnostics)
+            {
+                if (elementKeyValueTypes is (var elementKeyType, var elementValueType) &&
+                    elementConversion.TryGetKeyValueConversions(out var keyConversion, out var valueConversion))
```

Comment:
nit: consider inverting this `if` and the one below to flatten function a bit (if conversion fails, throw early; if we don't need an `Add` method, return early)

Reply:
It feels like this `if` condition would be harder to read if inverted. Will leave as is.